File: library/src/main/java/com/pengrad/telegrambot/request/AbstractMultipartRequest.java
package com.pengrad.telegrambot.request;

import java.io.File;

/**
 * stas
 * 5/1/16.
 */
abstract public class AbstractMultipartRequest<T extends AbstractMultipartRequest<T>> extends AbstractSendRequest<T> {

    private boolean isMultipart;
    private String fileName;
    private String contentType;

    public AbstractMultipartRequest(Object chatId, Object file) {
        super(chatId);
        if (file instanceof String) {
            isMultipart = false;
        } else if (file instanceof File) {
            isMultipart = true;
            fileName = ((File) file).getName();
        } else if (file instanceof byte[]) {
            isMultipart = true;
        } else {
            throw new IllegalArgumentException("Sending data should be String, File or byte[]");
        }
        add(getFileParamName(), file);
    }

    public T fileName(String fileName) {
        this.fileName = fileName;
        return thisAsT;
    }

    public T contentType(String contentType) {
        this.contentType = contentType;
        return thisAsT;
    }

    protected T thumbnail(Object thumbnail) {
        isMultipart = true;
        return add("thumbnail", thumbnail);
    }

    @Override
    public boolean isMultipart() {
        return isMultipart;
    }

    @Override
    public String getFileName() {
        if (fileName != null && !fileName.isEmpty()) {
            return fileName;
        } else {
            return getDefaultFileName();
        }
    }

    @Override
    public String getContentType() {
        return (contentType != null && !contentType.isEmpty()) ? contentType : getDefaultContentType();
    }

    abstract protected String getDefaultFileName();

    abstract protected String getDefaultContentType();

    abstract protected String getFileParamName();
}

File: library/src/main/java/com/pengrad/telegrambot/request/AbstractMultipartRequest.kt
package com.pengrad.telegrambot.request

import com.pengrad.telegrambot.utility.kotlin.optionalRequestParameter
import java.io.File

abstract class AbstractMultipartRequest<REQ : AbstractMultipartRequest<REQ>>(
    chatId: Long?,
    channelUsername: String?,

    contentParameterName: String,
    contentUrl: String?,
    contentFile: File?,
    contentBytes: ByteArray?,

    defaultFileName: String,
    defaultContentType: String
) : KAbstractSendRequest<REQ>(
    chatId = chatId,
    channelUsername = channelUsername
) {

    val contentUrl by optionalRequestParameter(contentUrl, customParameterName = contentParameterName)
    val contentFile by optionalRequestParameter(contentFile, customParameterName = contentParameterName)
    val contentBytes by optionalRequestParameter(contentBytes, customParameterName = contentParameterName)

    open val isMultipartRequest: Boolean
        get() = contentFile != null || contentBytes != null

    @get:JvmName("getContentFileName")
    var fileName: String = contentFile?.name ?: defaultFileName

    @get:JvmName("getContentMimeType")
    var contentType: String = defaultContentType

    fun fileName(fileName: String) = applySelf { this.fileName = fileName }

    fun contentType(contentType: String) = applySelf { this.contentType = contentType }

    override fun isMultipart() = isMultipartRequest

    override fun getFileName() = fileName

    override fun getContentType() = contentType

}